File: src/net/cyclestreets/liveride/HuntForSegment.java
package net.cyclestreets.liveride;

import net.cyclestreets.CycleStreetsPreferences;
import net.cyclestreets.routing.Journey;
import net.cyclestreets.routing.Segment;

import org.osmdroid.util.GeoPoint;

final class HuntForSegment extends LiveRideState
{
  private int waitToSettle_;
  
  HuntForSegment(final LiveRideState state) 
  {
    super(state);
    waitToSettle_ = 5;
  } // HuntForSegment
  
  @Override
  public LiveRideState update(Journey journey, GeoPoint whereIam, int accuracy)
  {
    if(waitToSettle_ > 0) {
      --waitToSettle_;
      return this;
    }
    
    Segment nearestSeg = null;
    int distance = Integer.MAX_VALUE;
    
    for(final Segment seg : journey.segments())
    {
      int from = seg.distanceFrom(whereIam);
      if(from < distance) 
      {
        distance = from;
        nearestSeg = seg;
      } // if ...
    } // for ...
    
    distance -= accuracy;

    if(distance > CycleStreetsPreferences.replanDistance())
      return new ReplanFromHere(this, whereIam);

    if(nearestSeg == journey.activeSegment())
      return new OnTheMove(this);
    
    return new AdvanceToSegment(this, journey, nearestSeg);
  } // update

  @Override
  public boolean isStopped() { return false; }
  @Override
  public boolean arePedalling() { return true; }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/HuntForSegment.kt
package net.cyclestreets.liveride

import net.cyclestreets.CycleStreetsPreferences
import net.cyclestreets.routing.Journey
import net.cyclestreets.routing.Segment
import org.osmdroid.util.GeoPoint

internal class HuntForSegment(state: LiveRideState) : LiveRideState(state) {

    private var waitToSettle = 5

    override fun update(journey: Journey, myLocation: GeoPoint, accuracy: Int): LiveRideState {
        if (waitToSettle > 0) {
            --waitToSettle
            return this
        }

        var nearestSeg: Segment = journey.segments.first()
        var distance = Int.MAX_VALUE

        for (seg in journey.segments) {
            val from = seg.distanceFrom(myLocation)
            if (from < distance) {
                distance = from
                nearestSeg = seg
            }
        }
        distance -= accuracy

        if (distance > CycleStreetsPreferences.replanDistance()) {
            return ReplanFromHere(this, myLocation)
        }
        if (nearestSeg === journey.activeSegment()) {
            return OnTheMove(this)
        }
        return AdvanceToSegment(this, journey, nearestSeg)
    }

    override fun isStopped(): Boolean { return false }
    override fun arePedalling(): Boolean { return true }
}
